xcube/core/store/preload.py

```diff
@@ -0,0 +1,445 @@
+# Copyright (c) 2018-2024 by xcube team and contributors
+# Permissions are hereby granted under the terms of the MIT License:
+# https://opensource.org/licenses/MIT.
+
+from abc import abstractmethod, ABC
+from asyncio import CancelledError
+from concurrent.futures import Executor, Future
+from concurrent.futures.thread import ThreadPoolExecutor
+from enum import Enum
+import threading
+from typing import Any, Callable
+
+import tabulate
+
+from xcube.util.assertions import assert_given, assert_instance
+
+
+class PreloadStatus(Enum):
+    """Preload process status."""
+
+    waiting = "waiting"
+    started = "started"
+    stopped = "stopped"
+    cancelled = "cancelled"
+    failed = "failed"
+
+    def __str__(self):
+        return self.name.upper()
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}.{self.name}"
+
+
+class PreloadState:
+    """Preload state."""
+
+    def __init__(
+        self,
+        data_id: str,
+        status: PreloadStatus | None = None,
+        progress: float | None = None,
+        message: str | None = None,
+        exception: BaseException | None = None,
+    ):
+        assert_given(data_id, name="data_id")
+        self.data_id = data_id
+        self.status = status
+        self.progress = progress
+        self.message = message
+        self.exception = exception
+
+    def update(self, event: "PreloadState"):
+        """Update this state with the given partial state.
+
+        Args:
+            event: the partial state.
+        """
+        assert_instance(event, PreloadState, name="data_id")
```

Comment:
I think the `name` in the `assert_instance` should be `event` so that the assertion reads `event must be an instance of PreloadState, was something_else`
```suggestion
        assert_instance(event, PreloadState, name="event")
```